design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,613 @@
+
+#### Flows
+
+The schema described in the logic would be implemented
+in the _Authentication > flows_ section. Some work can be done on the UX to make
+the process more intuitive, but this will be proposed at a later state in the
+design process.
+
+##### Default Flows
+
+Keycloak should have the same default flows as is currently the case, except that two
+authentication factors are used, with preset values for the authenticator executions:
+rather than adding and removing authenticator executions, the user can enable or disable
+them.
+This is only UX, because behind Keycloak sets or removes the **disabled** flag within the
+authentication execution's config. The 2nd authentication factor can also be disabled, and
+if either the 2nd factor is disabled, or if all authentication executions are disabled in
+the second factor, then the login will be single factor only.
+
+For example, for the browser flow this would be:
+```
+Auth type                         | Options
+---------------------------------------------------------------------------
+Identity Provider Redirector
+Authenticated on               
+  |- Cookie
+  |- Kerberos                      [ ] enabled  [x] disabled
+Authentication factor (1)
+  |- Username Password Form
+Authentication factor (2)          [x] enabled  [ ] disabled  [ ] optional
+  |- OTP Form                      [x] enabled  [ ] disabled
+  |- FIDO-2                        [x] enabled  [ ] disabled
+```
+And for the Direct Grant the flow would be for example:
+```
+Auth type                         | Options
+---------------------------------------------------------------------------
+Username Validation
+Authentication factor (1)
+  |- Password
+Authentication factor (2)          [x] enabled  [ ] disabled  [ ] optional
+  |- OTP                           [x] enabled  [ ] disabled
+  |- Other                         [x] enabled  [ ] disabled
+```
+
+##### Configurable flows
+
+Configurable flows work as is currently the case in Keycloak with the possibility
+to add new authentication flows and either replace the default flow in _Bindings_
+or in the _Client > [client] > Authentication Flow Overrides_. Here the normal configuration of
+an authentication factor is to add (or remove) authenticator executions. However, within an
+authentication execution's configuration, it is possible to set a **disable** flag.
+This is for when an admin doesn't want to keep an execution active, but doesn't want it's
+configuration to be deleted. An **optional** flag is present in the configuration of the
+authentication factor, meaning that it will only be required for users which have at least
+one credential set for an execution that is within that authentication factor.
+
+For UX purposes, the GUI shows directly in the flow when an authentication factor is set
+to optional, or if an authentication execution is disabled.
+
+#### Bindings
+
+No changes are necessary here.
+
+#### Required Actions
```

Comment:
I'm not following this section at all. I would say that what is needed is a required action to configure a valid 2nd factor, where the required action somehow lists what 2nd factor mechanisms are available and allows the user to pick. When the user picks it would somehow trigger the corresponding required action.

Reply:
Having a "set 2nd factor required action" was actually the initial suggestion that my colleagues had. However, we saw the following difficulties:
* How to find from the code the 2nd factor authentication categories to be set, when there is no explicit "2nd factor" type. And this is actually more difficult with the proposed subflow system, as there's even less to denote what a 2nd factor authentication is.
* Even if you _can_ distinguish all authentication categories that are a 2nd factor, what do you do when there are multiple flows defined, each with different authentication executions?

This is why we came up with the idea of configurable required actions, in which an administrator can explicitly say something like "I want the user to be required to set a credential for authentication category OTP or authentication category FIDO2". However, I suppose that we can simplify the configuration of such a required action to get the list of all authentication factors (or sub-flows) and the administrator simply chooses the one he wants. The required action can then infer the required authentication categories from the authentication executions within the selected flow.